marimo/_save/save.py

```diff
@@ -0,0 +1,161 @@
+
+        self._entered_trace = True
+        stack = traceback.extract_stack()
+        if not self._skipped:
+            return self._old_trace
+
+        # This only executes on the first line of code in the block. If the
+        # cache is hit, the block terminates early with a SkipWithBlock
+        # exception, if the block is not hit, self._skipped is set to False,
+        # causing this function to terminate before reaching this block.
+        for i, frame in enumerate(stack[::-1]):
+            _filename, lineno, function_name, _code = frame
+            if function_name == "<module>":
+                ctx = get_context()
+                assert (
+                    ctx.execution_context is not None
+                ), "Could not resolve context for cache."
+                graph = ctx.graph
+                cell_id = ctx.execution_context.cell_id
+                pre_module, save_module = ExtractWithBlock(lineno - 1).visit(
+                    ast.parse(ctx.graph.cells[cell_id].code).body
+                )
+
+                self._cache = hash_context(
+                    save_module,
+                    graph,
+                    cell_id,
+                    loader=self._loader,
+                    context=pre_module,
+                )
+                if self._cache.hit:
+                    raise SkipWithBlock()
+
+                self.cache_type = self._cache
+                self._skipped = False
+                return self._old_trace
+            elif i > 1:
+                raise CacheException(
+                    "persistent_cache must be invoked from cell level "
+                    "(cannot be in a function or class)"
+                )
+        raise CacheException("persistent_cache could not resolve block")
+
+    def __exit__(
+        self,
+        exception: Optional[Type[BaseException]],
+        instance: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ) -> bool:
+        sys.settrace(self._old_trace)  # Clear to previous set trace.
+        if not self._entered_trace:
+            raise CacheException("Unexpected block format.")
+
+        # Backfill the loaded values into global scope.
+        if self._cache and self._cache.hit:
+            for lookup, var in contextual_defs(self._cache):
+                globals()[var] = self._cache.defs[lookup]
+            # Return true to suppress the SkipWithBlock exception.
+            return True
+
+        # NB: exception is a type.
+        if exception:
+            assert not isinstance(
+                instance, SkipWithBlock
+            ), "Cache was not correctly set."
+            if isinstance(instance, BaseException):
+                raise instance
+            raise exception
+
+        # Fill the cache object and save.
+        assert self._cache is not None
+        for lookup, var in contextual_defs(self._cache):
+            self._cache.defs[lookup] = globals()[var]
+        self._loader.save_cache(self._cache)
```

Comment:
What's the behavior if the loader fails to serialize the object? For example not all objects are pickle-able.

Reply:
That's a good question. marimo doesn't really have a "warning", This could print to stderr that cache save did not occur. I don't think failure is completely correct.

Thoughts?

Reply:
Print to stdert sounds good to me.